### PR TITLE
Don't fail in `deleteIfManaged` when the namespace/project is not managed

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?

Don't fail in `deleteIfManaged` when the namespace/project is not managed.
Instead, ignore the request silently.

That seems to be the original intention based on the usage of the method
everywhere else but the test suite 8-|

This bug manifested itself as unnecessary warnings about the inability to delete the namespace/project when a workspace was deleted.

### What issues does this PR fix or reference?
#15686

### CODE REVIEW

**Comment by metlos** [comment]
ci-build

**Review by skabashnyuk — APPROVED** [review]

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
crw-ci-test

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 1]
```diff
@@ -141,23 +141,20 @@ void prepare(boolean markManaged, boolean canCreate) throws InfrastructureExcept
...
-      throw new InfrastructureException(
-          format(
-              "Can't delete namespace '%s' that contains"
-                  + " runtime of workspace '%s' because it doesn't have the '"
-                  + MANAGED_NAMESPACE_LABEL
-                  + "' label equal to 'true'.",
-              name,
-              workspaceId));
+    if (!isNamespaceManaged(client)) {
+      LOG.debug(
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: Maybe it makes sense to print `Log.info` instead. Then admin would understand better why namespace is not deleted after workspace creation.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java [hunk 2]
```diff
@@ -133,15 +134,12 @@ void deleteIfManaged() throws InfrastructureException {
...
-    if (!isManagedInternal(osClient)) {
-      throw new InfrastructureException(
-          format(
-              "Can't delete project '%s' that contains"
-                  + " runtime of workspace '%s' because it doesn't have the '"
-                  + MANAGED_NAMESPACE_LABEL
-                  + "' label equal to 'true'.",
-              projectName,
-              workspaceId));
+    if (!isProjectManaged(osClient)) {
```
Comment (sparkoo) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java:137]: I would negate the logic to more straight-forward:
```
if managed {
  delete
} else {
  LOG
}
```

**Review by sleshchenko — APPROVED** [review]
LGTM

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java [hunk 3]
```diff
@@ -133,15 +134,12 @@ void deleteIfManaged() throws InfrastructureException {
...
+      LOG.debug(
+          "Project {} for workspace {} is not marked as managed. Ignoring the delete request.",
+          projectName,
+          workspaceId);
+      return;
     }
 
     delete(projectName, osClient);
```
Comment (sparkoo) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java:145]: why you don't catch exception here, but you catch it in k8s class ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 4]
```diff
@@ -141,23 +141,20 @@ void prepare(boolean markManaged, boolean canCreate) throws InfrastructureExcept
...
+          "Namespace {} for workspace {} is not marked as managed. Ignoring the delete request.",
+          name,
+          workspaceId);
+      return;
     }
 
     try {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:160]: can we catch delete exception with 403 check in delete method?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 5]
```diff
@@ -141,23 +141,20 @@ void prepare(boolean markManaged, boolean canCreate) throws InfrastructureExcept
...
-    if (!isManagedInternal(client)) {
-      throw new InfrastructureException(
-          format(
-              "Can't delete namespace '%s' that contains"
-                  + " runtime of workspace '%s' because it doesn't have the '"
-                  + MANAGED_NAMESPACE_LABEL
-                  + "' label equal to 'true'.",
-              name,
-              workspaceId));
+    if (!isNamespaceManaged(client)) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:152]: same as OS, I would negate the logic `if managed -> delete`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 6]
```diff
@@ -141,23 +141,20 @@ void prepare(boolean markManaged, boolean canCreate) throws InfrastructureExcept
 
   /**
    * Deletes the namespace. Deleting a non-existent namespace is not an error as is not an attempt
-   * to delete a namespace that is already being deleted.
+   * to delete a namespace that is already being deleted. If the namespace is not marked as managed,
+   * it is silently not deleted.
    *
-   * @throws InfrastructureException if the namespace is not marked managed or when any unexpected
-   *     exception occurs during namespace deletion
+   * @throws InfrastructureException f any unexpected exception occurs during namespace deletion
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: `InfrastructureException f any...` - `f` should probably be `if`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 5 above)
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:152]: I don't understand this part of code from `isNamespaceManaged()`

```
} else if (e.getCode() == 404) {
        // we don't want to block whatever work the caller is doing on the namespace. The caller
        // will fail anyway if the namespace doesn't exist.
        return true;
      }
```
It does not make logical sense to me when something does not exist that I manage it.

**Review by sparkoo — COMMENTED** [review]
Few comments a bit outside from your changes. However, kubernetes and OS logic is the same, but code is a bit off.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java (hunk 2 above)
Comment (metlos) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java:137]: I prefer less nesting. The `if (!is*Managed()` can be thought of as a precondition to the actual delete.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 7]
```diff
@@ -141,23 +141,20 @@ void prepare(boolean markManaged, boolean canCreate) throws InfrastructureExcept
 
   /**
    * Deletes the namespace. Deleting a non-existent namespace is not an error as is not an attempt
-   * to delete a namespace that is already being deleted.
+   * to delete a namespace that is already being deleted. If the namespace is not marked as managed,
+   * it is silently not deleted.
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:145]: I don't know if silently defaulting to a no-op is the best choice, something should likely be logged.

**Review by amisevsk — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 7 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:145]: Something is logged :)

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 1 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: Good idea. Done.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java (hunk 3 above)
Comment (metlos) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java:145]: Let's solve this under https://github.com/eclipse/che/issues/15250

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 4 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:160]: Let's solve this under https://github.com/eclipse/che/issues/15250

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 5 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:152]: Let's solve this under #15250

**Comment by metlos** [comment]
> Few comments a bit outside from your changes. However, kubernetes and OS logic is the same, but code is a bit off.

I agree to the degree that I already filed an issue a couple of months ago :) - #15250

I think we should not embark on such a potentially big refactoring in a small bugfix like this.

**Comment by metlos** [comment]
crw-ci-test

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 7 above)
Comment (amisevsk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:145]: In that case, "silently" is a bit misleading ;)

**Pull request merged**